src/controllers/userController.js

```diff
@@ -0,0 +1,163 @@
+    const { uid } = req.params;
+
+    try {
+        const userRecord = await admin.auth().getUser(uid);
+        res.status(200).send(userRecord);
+    } catch (error) {
+        console.error('Error fetching user:', error);
+        res.status(400).send({ error: error.message });
+    }
+};
+
+exports.getAllUsers = async (req, res) => {
+    try {
+        const usersSnapshot = await db.collection('users').get();
+        if (usersSnapshot.empty) {
+            return res.status(404).send({ message: 'No users found' });
+        }
+        
+        const users = [];
+        usersSnapshot.forEach(doc => {
+            users.push({ id: doc.id, ...doc.data() });
+        });
+
+        res.status(200).send(users);
+    } catch (error) {
+        console.error('Error fetching users:', error);
+        res.status(500).send({ error: error.message });
+    }
+};
+
+exports.logoffUser = async (req, res) => {
+    try {
+        const { uid } = req.params;
+        if (!uid) {
+            return res.status(400).send({ error: 'User ID is required' });
+        }
+
+        await admin.auth().revokeRefreshTokens(uid);
+
+        res.status(200).send({ message: 'User logged off successfully' });
+    } catch (error) {
+        console.error('Error logging off user:', error);
+        res.status(400).send({ error: error.message });
+    }
+};
+
+exports.resetPassword = async (req, res) => {
+    const { email } = req.body;
+
+    if (!email) {
+        return res.status(400).send({ error: 'Email is required' });
+    }
+
+    try {
+        const apiKey = process.env.FIREBASE_API_KEY;
+
+        await axios.post(`https://identitytoolkit.googleapis.com/v1/accounts:sendOobCode?key=${apiKey}`, {
+            requestType: 'PASSWORD_RESET',
+            email
+        });
+
+        res.status(200).send({ message: 'Password reset email sent' });
+    } catch (error) {
+        console.error('Error sending password reset email:', error);
+        if (error.response) {
+            console.error('Error response data:', error.response.data);
+        }
+        res.status(400).send({ error: error.message });
+    }
+};
+
+
+exports.updateUserProfile = async (req, res) => {
+    const { uid } = req.params; // User ID from URL params
+    const { displayName, photoURL } = req.body; // New name and profile picture URL
+
+    if (!uid || (!displayName && !photoURL)) {
+        return res.status(400).json({ error: 'User ID and at least one field to update are required' });
+    }
+
+    try {
+        const updates = {};
+        if (displayName) updates.displayName = displayName;
+        if (photoURL) updates.photoURL = photoURL;
+
+        const userRecord = await admin.auth().updateUser(uid, updates);
+        
+        res.status(200).send({ message: 'User profile updated successfully', user: userRecord });
+    } catch (error) {
+        console.error('Error updating user profile:', error);
+        res.status(400).send({ error: error.message });
```

Comment:
This should be a 500 error (internal server error). 400 indicates the request object had invalid data and it's the client's fault, 500 indicates the server tried but something went wrong internally.